### PR TITLE
Extensions to beam_decode for random sampling

### PR DESCRIPTION
**Pull request by davidweichiang**
It looks to me like it's working, but @tnq177 if you can take a look to make sure that (1) it looks right and (2) you're okay with adding some clutter to beam_decode, it would be appreciated!

### CODE REVIEW

File: layers/encoders.py [hunk 1]
```diff
@@ -194,8 +198,14 @@ def beam_decode(self, encoder_out, encoder_mask, get_input_fn, logprob_fn, bos_i
 
         # Length penalty not needed, because all lengths are 1
 
-        # Select top k hyps to survive to the next time step
-        all_probs, symbols = torch.topk(probs, beam_size, dim=-1) # ([bsz, beam], [bsz, beam])
+        if mode == "best":
+            # Select top k hyps to survive to the next time step
+            all_probs, symbols = torch.topk(probs, beam_size, dim=-1) # ([bsz, beam], [bsz, beam])
+        elif mode == "random":
+            symbols = torch.multinomial(torch.exp(probs), beam_size, replacement=True)
```
Comment (tnq177) [inline layers/encoders.py:206]: I think it's better to calculate probs from logits rather than taking exp of logprobs here.

**Comment by tnq177** [comment]
lgtm except for the comment i made here https://github.com/tnq177/witwicky/pull/6/commits/3d1548b54df1484292539e327a14c11c81614046#r297830997

File: layers/encoders.py (hunk 1 above)
Comment (davidweichiang) [reply layers/encoders.py:206]: gp, will fix

Comment (davidweichiang) [reply layers/encoders.py:206]: oh, but what if logprob_fn is something other than log_softmax?

Comment (tnq177) [reply layers/encoders.py:206]: hmm why would we need a logprob_fn other than log_softmax?

Comment (davidweichiang) [reply layers/encoders.py:206]: I don't know...but you made `logprob_fn` an argument to `beam_decode`, so it seems like in principle it could be something else...

At any rate, I am working on an update to this PR but haven't gotten it right yet. So please don't merge.
